sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RegionScopedSessionContainer.java

```diff
@@ -0,0 +1,621 @@
+                    this.collectionResourceIdToCollectionName.compute(resourceId.getUniqueDocumentCollectionId(), (k, v) -> collectionName);
+                }
+                this.addSessionTokenAndTryRecordEpkInBloomFilter(request, resourceId, partitionKeyRangeId, parsedSessionToken);
+            } finally {
+                this.writeLock.unlock();
+            }
+        }
+    }
+
+    private void recordPartitionKeyInBloomFilter(
+        RxDocumentServiceRequest request,
+        Long collectionRid,
+        String regionRoutedTo,
+        PartitionKeyInternal partitionKeyInternal,
+        PartitionKeyDefinition partitionKeyDefinition) {
+
+        if (Strings.isNullOrEmpty(this.firstPreferredReadableRegionCached.get())) {
+            this.firstPreferredReadableRegionCached.set(extractFirstEffectivePreferredReadableRegion());
+        }
+
+        this.partitionKeyBasedBloomFilter.tryRecordPartitionKey(
+            request,
+            collectionRid,
+            this.firstPreferredReadableRegionCached.get(),
+            regionRoutedTo,
+            partitionKeyInternal,
+            partitionKeyDefinition);
+    }
+
+    private void recordRegionScopedSessionToken(
+        RxDocumentServiceRequest request,
+        PartitionScopedRegionLevelProgress partitionScopedRegionLevelProgress,
+        ISessionToken parsedSessionToken,
+        String partitionKeyRangeId,
+        String regionRoutedTo) {
+
+        partitionScopedRegionLevelProgress.tryRecordSessionToken(
+            request,
+            parsedSessionToken,
+            partitionKeyRangeId,
+            this.firstPreferredReadableRegionCached.get(),
+            regionRoutedTo);
+    }
+
+    private void addSessionTokenAndTryRecordEpkInBloomFilter(RxDocumentServiceRequest request, ResourceId resourceId, String partitionKeyRangeId, ISessionToken parsedSessionToken) {
+
+        final Long collectionResourceId = resourceId.getUniqueDocumentCollectionId();
+
+        PartitionScopedRegionLevelProgress partitionScopedRegionLevelProgress
+            = this.collectionResourceIdToPartitionScopedRegionLevelProgress.get(collectionResourceId);
+
+        if (this.firstPreferredReadableRegionCached.get().equals(StringUtils.EMPTY)) {
+            this.firstPreferredReadableRegionCached.set(extractFirstEffectivePreferredReadableRegion());
+        }
+
+        String regionRoutedTo = null;
+
+        if (request.requestContext != null) {
+            URI regionEndpointRoutedTo = request.requestContext.locationEndpointToRoute;
+            regionRoutedTo = this.globalEndpointManager.getRegionName(regionEndpointRoutedTo, request.getOperationType());
+        }
+
+        Utils.ValueHolder<PartitionKeyInternal> partitionKeyInternal = Utils.ValueHolder.initialize(null);
+        Utils.ValueHolder<PartitionKeyDefinition> partitionKeyDefinition = Utils.ValueHolder.initialize(null);
+
+        if (partitionScopedRegionLevelProgress != null) {
+
+            if (shouldUseBloomFilter(
```

Comment:
as discussed offline, I am wondering when such case happens (like cross partition query happens), should we disable the optimization altogether (to guarantee read after read etc)

Reply:
Fixed in next iteration(s).